### PR TITLE
Add method to reset use_color setting.

### PR DESCRIPTION
**Pull request by phiggins**
We ran into some order dependent test failures because some tests were setting `HighLine.use_color` and not resetting it in between test runs. When we figured out the problem, we reset the value in a before block for our rspec suite. Since we need to change HighLine's color setting and there was no easy way to reset it, it seemed like a good addition to the `reset` method.

Before we had to do something like this:

```
before do
  @old_highline_use_color = HighLine.use_color?
  HighLine.use_color = true
end

after do
  HighLine.use_color = @old_highline_use_color
end
```

After this PR we would only have to do:

```
before do
  HighLine.reset
end
```

### CODE REVIEW

**Comment by phiggins** [comment]
I just noticed that master is a prerelease version of the library. I'd be happy to redo this PR against the 1.7 branch if that would make it more likely to get into a release!

**Comment by abinoam** [comment]
Hi @phiggins, are you using stable or prerelease version at your production software?
I've postponed 2.0.0 release for too long, I guess.
I'll try to make it this or next month.
About your PR, as soon as possible I'll take a look at it and try to merge it.
Thank you very much for contributing to HighLine development!

**Comment by abinoam** [comment]
Hi @phiggins,

Perhaps the code at the PR doesn't fit your requirements.
As far as I could understand, you want `reset_color` to go back to a previous `use_color` value. Am I understanding it correctly?

This is what I see at the snippet you have provided.
```ruby
before do
  @old_highline_use_color = HighLine.use_color?
  HighLine.use_color = true
end

after do
  HighLine.use_color = @old_highline_use_color
end
```

Above you save the current `use_color` value and restore it later.

But, at your PR, you just set it to `true`.

```ruby
# Resets the use of color.
def self.reset_use_color
  @use_color = true
end
```

Please, correct me if I have missed something.

**Comment by abinoam** [comment]
And... perhaps saving its previous value shouldn't be HighLine's responsability. I think your snippet (where you save the value to restore it later) works well and perhaps is the right way to do it. I mean, perhaps we shouldn't patch Highline to add such functionality. What do you think about it?

**Comment by abinoam** [comment]
Hi @phiggins, rereading your PR I think you are pointing me toward to a bigger issue.
This `use_color` switch is being saved as (semi) global state, which is bad and we've been trying to change some of them at 2.0.0 branch.

We were a little conservative at 2.0.0 as HighLine is a pretty mature library and there's a lot of software that depends on it. So, completely breaking the api at this first step (v.2.0.0) would not be desirable. So we have tried to maintain maximum compatibility with stable branch.

I think we could address this by completly changing the way we save this (and other switches/options).

Just as example, we have changed the recomendation.
Before 2.0.0 we were doing things like this
```ruby
# Poluting the global namespace with #say, #ask and other methods
require 'highline/import'

ask "What do you think about this?"
```

Now we recomend avoiding polluting global namespace
```ruby
require 'highline'

cli = HighLine.new
cli.ask "Much better, isn't it?"
```

**Comment by abinoam** [comment]
Hi @phiggins, could you have a look at #212 ?

**Comment by phiggins** [comment]
Hi, thanks for your quick response!

> As far as I could understand, you want reset_color to go back to a previous use_color value. > Am I understanding it correctly?

I think I was not clear exactly what we're trying to do. Somewhere in our implementation code we're doing something like this `HighLine.use_color = some_condition?` to mutate HighLine's global state. In our tests, we want to be able to reset HighLine's state to some known value so that global state does not persist between our tests, and this PR was an attempt to make that more straightforward.

This PR is not an attempt to get HighLine to remember a previous setting, it is an attempt to give HighLine the ability to be reset to its default value.

I think what you've said here and in #212 is probably the best way forward for new code. Unfortunately we have several repos that depend on HighLine having global settings and it's unlikely that it'll ever be a priority to update them.

**Comment by abinoam** [comment]
Hi @phiggins, It's more clear to me now.

I'll cherry pick your commit and improve on PR #212 .

I think I can solve both issues altogether.

**Comment by abinoam** [comment]
I have cherry-picked your commit here https://github.com/JEG2/highline/pull/212/commits/09bf55b401ee1d6082f15dffc8b460d7c7c71732 
I have also added a test to cover one of your additions (HighLine.reset calling HighLine.reset_use_color) -> https://github.com/JEG2/highline/pull/212/commits/e54a06cd3039f8654fd4dbfc0083b62dc1b852f8 (Just copied your added test and changed it to call HighLine.reset).

Now I think it will work as you expect. Please, tell me if I get something wrong.

About the global state thing. 
Basically:
1) Requiring `'highline/import'` creates a `HighLine` instance and saves it at the global variable `$terminal`
2) Some of the `HighLine` global state were being saved as class instance variables at `HighLine` itself.
3) What I'm trying to do is to push everything to the instance. And add delegators to `$terminal`, so I can unify the api while making it backward compatible.

Thank you for your PR and conversation! :+1:
I'll merge PR #212 and ask @JEG2 to release it.

**Comment by abinoam** [comment]
Closing this PR as the commit was cherry-picked into PR #212 (https://github.com/JEG2/highline/commit/09bf55b401ee1d6082f15dffc8b460d7c7c71732)

**Comment by phiggins** [comment]
Hi,

I looked over the code in #212 and I think it looks good. Thanks for taking the idea and polishing it up!

I hadn't seen your use of `$terminal`, but I had been thinking of how to move to a more stateless approach while keeping backwards compatibility. The solution I had in mind was to implement a singleton that kept the same interface as the old stateful stuff but delegated to an instance underneath. Here is a sketch to try to explain:

```
class HighLine
  class Legacy
    
    # Setup the singleton to always have an instance it delegates to that can be reset easily.
    def self.instance
      @instance ||= HighLine.new
    end

    def self.reset!
      @instance = nil
    end

    # Delegate the old class-level stateful methods to the instance.
    def self.use_color=(setting)
      instance.use_color = setting
    end
end
```

I'm glad that you're working on this stuff and I look forward to seeing it get released!

**Comment by abinoam** [comment]
Hi @phiggins, 

Thank you for the ideas. I think they will be useful as soon as I begin to deprecate any incompatible api.

The `$terminal` thing is on https://github.com/JEG2/highline/blob/master/lib/highline/import.rb
We're already doing something similar to what you suggested. 
But, the use of this `$terminal` has already brought us incompatibilities with other gems that tried to do the same. So I think encapsulating this "global" state under the HighLine class (as a namespace) as you suggested will be welcome.
